articles/introduction-to-avl-trees/index.md

```diff
@@ -0,0 +1,548 @@
+
+### Reason for AVL Trees
+
+Operations on BSTs like search, insertion and deletion take O(h) time, where 'h' represents the height of the tree. In case of a skewed BST (a binary search tree where all nodes have only one or no child node), performing these operations take O(n) time, where 'n' denotes the number of nodes in the tree. To tackle this inefficiency in the process, the tree can be constructed in such a way that it always maintains a logarithmic height, thereby reducing the time complexity for all operations to O(log h).
+
+This means that the height of the tree must be maintained after every insertion and deletion such that the time complexity for every operation performed thereafter remains constant.
+
+How does one maintain the height of the tree after every such modification?
+
+### Tri-Node Restructuring
+When a node is either inserted or deleted in an AVL tree, it might create an imbalance in the tree and would result in an increase in the time complexity. But, if the tree is restructured every time there is a modification to the structure, it would ensure that the time complexity still remains O(log h).
+
+Let's see how the insertion of a node into the following tree would affect its structure.
+
+Take the example of the following AVL Tree. 
+
+<center> <img src="/engineering-education/introduction-to-avl-trees/avl_insertion_before.jpg" alt="Before insertion"> </center>
+
+On inserting a node (4) in the tree, the structure would change to -
+
+<center> <img src="/engineering-education/introduction-to-avl-trees/tree_after_insertion.jpg" alt="After Insertion"> </center>
+
+In tri-node restructuring, the three roots which are involved in the imbalanced structure are chosen and are rearranged such that the order is restored.
+In this case, the roots of node (3), node (4), and node(5) are taken into account because the imbalance occurs at node (4). If the height at node (4) is balanced, the overall height of the tree would remain logarithmic. For example, node (4) could be restructured similar to the figure below.
+
+<center> <img src="/engineering-education/introduction-to-avl-trees/tree_after_structuring.jpg" alt="After Tri-Node Restructuring"> </center>
+
+This ensures that the height of all the nodes is balanced and hence, an AVL Tree. 
+
+#### Tri-Node Restructuring for Insertion
+
+The procedure for tri-node restructuring for an insertion operation is as follows - 
+
+    1. Perform regular BST insertion.
+    2. Let the node being inserted be known as 'a'. Let 'd' be the first node going up from 'w' towards root that is unbalanced. Let 'c' be a child of 'd' that lies on the path between 'a' and 'd'. Let 'b' be a child of 'c' that comes on the same path. 
+    3. This would result in four possible cases:
+        - Left-left rotation: Where 'c' is the left child of 'd' and 'b' is the left child of 'c' 
```

Comment:
> Left-left rotation: Where 'c' is the left child of 'd' and 'b' is the left child of 'c'

This line would look more readable if you could add bold formatting to these headings.

```md
**Left-Left Rotation**
```

Accordingly, do the same for all rotation types that you've mentioned in the article.

Reply:
Revisions made in the recent commit.